server/client.go

```diff
@@ -0,0 +1,385 @@
+	"net/http"
+	"net/url"
+
+	jira "github.com/andygrunwald/go-jira"
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-server/plugin"
+)
+
+// Client is the combined interface for all upstream APIs and convenience methods.
+type Client interface {
+	RESTService
+	IssueService
+	ProjectService
+	SearchService
+	UserService
+}
+
+// RESTService is the low-level interface for invoking the upstream service.
+// endoint can be a "short" API URL path, including the version desired, like "3/user",
+// or a fully-qualified URL, with a non-empty Scheme.
+type RESTService interface {
+	RESTGet(endpoint string, params map[string]string, dest interface{}) error
+	RESTPostAttachment(issueID string, data []byte, name string) (*jira.Attachment, error)
+}
+
+// UserService is the interface for user-related APIs.
+type UserService interface {
+	GetSelf() (*jira.User, error)
+	GetUserGroups(user JIRAUser) ([]*jira.UserGroup, error)
+}
+
+// ProjectService is the interface for project-related APIs.
+type ProjectService interface {
+	GetProject(key string) (*jira.Project, error)
+}
+
+// SearchService is the interface for search-related APIs.
+type SearchService interface {
+	SearchIssues(jql string, options *jira.SearchOptions) ([]jira.Issue, error)
+	SearchUsersAssignableToIssue(issueKey, query string, maxResults int) ([]jira.User, error)
+}
+
+// IssueService is the interface for issue-related APIs.
+type IssueService interface {
+	GetIssue(key string, options *jira.GetQueryOptions) (*jira.Issue, error)
+	CreateIssue(issue *jira.Issue) (*jira.Issue, error)
+
+	AddAttachment(api plugin.API, issueKey, fileID string, maxSize ByteSize) (mattermostName, jiraName string, err error)
+	AddComment(issueKey string, comment *jira.Comment) (*jira.Comment, error)
+	DoTransition(issueKey, transitionID string) error
+	GetCreateMeta(*jira.GetQueryOptions) (*jira.CreateMetaInfo, error)
+	GetTransitions(issueKey string) ([]jira.Transition, error)
+	UpdateAssignee(issueKey string, user *jira.User) error
+	UpdateComment(issueKey string, comment *jira.Comment) (*jira.Comment, error)
+}
+
+// JiraClient is the common implementation of most Jira APIs, except those that are
+// Jira Server or Jira Cloud specific.
+type JiraClient struct {
+	Jira *jira.Client
+}
+
+// RESTGet calls a specified HTTP point with a GET method. endpoint must be an absolute URL, or a
+// relative URL starting with a version, like "2/user".
+func (client JiraClient) RESTGet(endpoint string, params map[string]string, dest interface{}) error {
+	endpointURL, err := restURL(endpoint)
+	if err != nil {
+		return err
+	}
+	req, err := client.Jira.NewRequest("GET", endpointURL, nil)
+	if err != nil {
+		return err
+	}
+
+	q := req.URL.Query()
+	for k, v := range params {
+		q.Add(k, v)
+	}
+	req.URL.RawQuery = q.Encode()
+
+	resp, err := client.Jira.Do(req, dest)
+	if err != nil {
+		err = userFriendlyJiraError(resp, err)
+	}
+	return err
+}
+
+// RESTPostAttachment uploads an attachment to an issue
```

Comment:
Should probably comment, this is the gist of the bug - the issue/attachments API returns _some_ errors in plain text, and some in JSON. This is a good possibility to contribute back to go-jira, too.